### PR TITLE
rework ModelMethodOrder linter

### LINKED ISSUE #162: `ModelMethodOrder` output isn't helpful
The output provided by the `ModelMethodOrder` lint isn't helpful. Below is my `User` model which the linter complains about but so far I got the method order is right!?

* relationships: there are none
* scopes: there are none
* accessors: first two method `getFullNameAttribute` and `getAvatarAttribute`
* mutators: third method `setPasswordAttribute`
* boot: there isn't one
* everything else is placed at the end

I would expect the lint the provide the order it expects the methods to be. 

```bash
$ tlint lint app

Lints for /Users/tomwitkowski/git/Gummibeer/shootager/app/Models/User.php
============
! Model method order should be relationships > scopes > accessors > mutators > boot
55 : `class User extends Model implements AuthenticatableContract, AuthorizableContract, CanResetPasswordContract, MustVerifyEmailContract, GetsCleanedUp`
```

<details>
  <summary>app/Models/User.php</summary>

```php
<?php

namespace App\Models;

use App\Notifications\ResetPassword as ResetPasswordNotification;
use App\Notifications\VerifyEmail as VerifyEmailNotification;
use Astrotomic\Unavatar\Unavatar;
use Illuminate\Auth\Authenticatable;
use Illuminate\Auth\MustVerifyEmail;
use Illuminate\Auth\Passwords\CanResetPassword;
use Illuminate\Contracts\Auth\Access\Authorizable as AuthorizableContract;
use Illuminate\Contracts\Auth\Authenticatable as AuthenticatableContract;
use Illuminate\Contracts\Auth\CanResetPassword as CanResetPasswordContract;
use Illuminate\Contracts\Auth\MustVerifyEmail as MustVerifyEmailContract;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Foundation\Auth\Access\Authorizable;
use Illuminate\Notifications\DatabaseNotification;
use Illuminate\Notifications\DatabaseNotificationCollection;
use Illuminate\Notifications\Notifiable;
use Illuminate\Support\Carbon;
use Illuminate\Support\Facades\Hash;
use Spatie\ModelCleanup\CleanupConfig;
use Spatie\ModelCleanup\GetsCleanedUp;

/**
 * App\Models\User.
 *
 * @property int $id
 * @property string $first_name
 * @property string $last_name
 * @property string|null $business_name
 * @property string $email
 * @property string $password
 * @property string|null $remember_token
 * @property Carbon|null $email_verified_at
 * @property Carbon|null $created_at
 * @property Carbon|null $updated_at
 * @property-read string $full_name
 * @property-read string $avatar
 * @property-read DatabaseNotificationCollection|DatabaseNotification[] $notifications
 * @property-read int|null $notifications_count
 *
 * @method static Builder|User newModelQuery()
 * @method static Builder|User newQuery()
 * @method static Builder|User query()
 * @method static Builder|User whereBusinessName($value)
 * @method static Builder|User whereCreatedAt($value)
 * @method static Builder|User whereEmail($value)
 * @method static Builder|User whereEmailVerifiedAt($value)
 * @method static Builder|User whereFirstName($value)
 * @method static Builder|User whereId($value)
 * @method static Builder|User whereLastName($value)
 * @method static Builder|User wherePassword($value)
 * @method static Builder|User whereRememberToken($value)
 * @method static Builder|User whereUpdatedAt($value)
 * @mixin \Eloquent
 */
class User extends Model implements AuthenticatableContract, AuthorizableContract, CanResetPasswordContract, MustVerifyEmailContract, GetsCleanedUp
{
    use Notifiable, Authenticatable, Authorizable, CanResetPassword, MustVerifyEmail;

    protected $fillable = [
        'first_name',
        'last_name',
        'business_name',
        'email',
        'password',
    ];

    protected $hidden = [
        'password',
        'remember_token',
    ];

    protected $dates = [
        'email_verified_at',
    ];

    protected $casts = [];

    public function getFullNameAttribute(): string
    {
        return sprintf('%s %s', $this->first_name, $this->last_name);
    }

    public function getAvatarAttribute(): string
    {
        return Unavatar::email($this->email)->toUrl();
    }

    public function setPasswordAttribute(string $value): void
    {
        if (Hash::info($value)['algo'] === null) {
            $value = Hash::make($value);
        }

        $this->attributes['password'] = $value;
    }

    public function cleanUp(CleanupConfig $config): void
    {
        $config
            ->olderThanDays(7)
            ->scope(fn (Builder $query) => $query->whereNull('email_verified_at'));
    }

    public function sendEmailVerificationNotification(): void
    {
        $this->notify(new VerifyEmailNotification($this));
    }

    public function sendPasswordResetNotification($token): void
    {
        $this->notify(new ResetPasswordNotification($this, $token));
    }
}
```

</details>

### PR DESCRIPTION
**Pull request by Gummibeer**
fixes #162 

The linter now also knows the `booting()` and `booted()` methods and can differentiate `relationship` and `custom` methods by return type.
The description now also contains an explicit listing of all model methods with the matched type and ordered similar to what the linter expects.

### CODE REVIEW

File: tests/Linting/Linters/ModelMethodOrderTest.php [hunk 1]
```diff
@@ -11,46 +11,104 @@ class ModelMethodOrderTest extends TestCase
...
-    public function phone()
+
+    public function phone(): HasOne
     {
         return \$this->hasOne('App\Phone');
     }
+
+    public function tags(): HasMany
+    {
+        return \$this->hasMany('App\Tag');
```
Comment (loganhenson) [inline tests/Linting/Linters/ModelMethodOrderTest.php:79]: Could we potentially match on the common relationship method names (e.g. `hasMany` etc) as a best effort match strategy?

Comment (Gummibeer) [reply tests/Linting/Linters/ModelMethodOrderTest.php:79]: How about both!?
But yeah, will try to make this work. Have dig into the parser lib. 😅

Comment (loganhenson) [reply tests/Linting/Linters/ModelMethodOrderTest.php:79]: Oh definitely both if possible haha, but yea it is a bit trickier 😅

Comment (Gummibeer) [reply tests/Linting/Linters/ModelMethodOrderTest.php:95]: https://github.com/tightenco/tlint/pull/163/commits/3b1940a01eab268bfae14d38f93e405ad292f666

Have also made the `boot` ones stricter and added a `custom` test that flags every `abstract`, `private`, `protected` or `static` method as `custom`. This way the other tests don't have to test for method visibility or anything.
The `$this->tests` declaration is also commented that the order is important and comments also every of the 3 blocks. Have also added some more methods to the test that should be ordered the right way.

**Comment by loganhenson** [comment]
@Gummibeer I'll exercise this branch on some codebases and try to get it merged/released Friday, thank you so much for the contribution!

**Comment by Gummibeer** [comment]
I think the best change is that thee linter tells the user how each method is detected and also an optional sorting. If I would find a way to move whole methods with everything related it would even be possible to add a fixer. 🤔

**Pull request merged**